concordium-consensus/src/Concordium/GlobalState/Persistent/BlockState/Updates.hs

```diff
@@ -605,6 +634,65 @@ processBakerStakeThresholdUpdates t bu = do
                     pendingUpdates = pendingUpdates{pBakerStakeThresholdQueue = newpQ}
                 }
 
+-- |Process the add anonymity revoker update queue.
+--  Ignores updates with duplicate ARs.
+processAddAnonymityRevokerUpdates :: (MonadBlobStore m)
+  => Timestamp
+  -> BufferedRef Updates
+  -> HashedBufferedRef ARS.AnonymityRevokers
+  -> m (Map.Map TransactionTime UpdateValue, BufferedRef Updates, HashedBufferedRef ARS.AnonymityRevokers)
+processAddAnonymityRevokerUpdates t bu hbar = do
+    u@Updates{..} <- refLoad bu
+    oldARs <- ARS.arRevokers <$> refLoad hbar
+    oldQ <- refLoad (pAddAnonymityRevokerQueue pendingUpdates)
+    let (ql, qr) = Seq.spanl ((<= t) . transactionTimeToTimestamp . fst) (uqQueue oldQ)
+    (changes, updatedARs) <- addAndAccumNonduplicateUpdates oldARs ARS.arIdentity UVAddAnonymityRevoker ql
+
+    newQ <- refMake oldQ { uqQueue = qr }
+    newU <- refMake u { pendingUpdates = pendingUpdates {pAddAnonymityRevokerQueue = newQ} }
+    updatedARsRef <- refMake . ARS.AnonymityRevokers $ updatedARs
+    return (changes, newU, updatedARsRef)
+
+-- |Process the add identity provider update queue.
+--  Ignores updates with duplicate IPs.
+processAddIdentityProviderUpdates :: (MonadBlobStore m)
+  => Timestamp
+  -> BufferedRef Updates
+  -> HashedBufferedRef IPS.IdentityProviders
+  -> m (Map.Map TransactionTime UpdateValue, BufferedRef Updates, HashedBufferedRef IPS.IdentityProviders)
+processAddIdentityProviderUpdates t bu hbip = do
+    u@Updates{..} <- refLoad bu
+    oldIPs <- IPS.idProviders <$> refLoad hbip
+    oldQ <- refLoad (pAddIdentityProviderQueue pendingUpdates)
+    let (ql, qr) = Seq.spanl ((<= t) . transactionTimeToTimestamp . fst) (uqQueue oldQ)
+    (changes, updatedIPs) <- addAndAccumNonduplicateUpdates oldIPs IPS.ipIdentity UVAddIdentityProvider ql
+
+    newQ <- refMake oldQ { uqQueue = qr }
+    newU <- refMake u { pendingUpdates = pendingUpdates {pAddIdentityProviderQueue = newQ} }
+    updatedIPsRef <- refMake . IPS.IdentityProviders $ updatedIPs
+    return (changes, newU, updatedIPsRef)
+
+-- |Used for adding new IPs and ARs.
+-- Ensuring that new IPs/ARs have unique ids is difficult when enqueueing.
+-- Instead, it is handled here by ignoring updates with duplicate IPs/ARs.
+-- It also accumulates the actual changes that occured.
+addAndAccumNonduplicateUpdates :: (MonadBlobStore m, Foldable f, Reference m ref (StoreSerialized v), Ord k)
+                      => StrictMap.Map k v -- ^ The existing IPs / ARs.
+                      -> (v -> k) -- ^ Getter for the key field.
+                      -> (v -> UpdateValue) -- ^ Data constructor for UpdateValue.
+                      -> f (TransactionTime, ref (StoreSerialized v)) -- ^ The updates.
+                      -> m (Map.Map TransactionTime UpdateValue, StrictMap.Map k v)
+addAndAccumNonduplicateUpdates oldMap getKey toUV = foldM go (Map.empty, oldMap)
```

Comment:
The use of foldM here is inconsistent with the use of `foldr` in the basic block state case. They traverse the list differently so they will disagree in cases where there are conflicting updates.